File: packages/flutter-dart/lib/appkit_auth/social_login.dart
import 'package:flutter/services.dart';
import 'models/auth_result.dart';
import 'models/auth_error.dart';

/// Social login manager for Flutter
class SocialLoginManager {
  final String authUrl;
  final String projectId;

  SocialLoginManager({
    this.authUrl = 'https://auth.cinacoin.com',
    required this.projectId,
  });

  /// Sign in with Google (via url_launcher + deep link)
  Future<AuthResult> signInWithGoogle() async {
    final callbackUrl = '$authUrl/api/auth/google/callback';
    final authUrlStr =
        '$authUrl/api/auth/google?redirect_uri=$callbackUrl&project_id=$projectId';

    // In production, use url_launcher to open browser and handle deep link callback
    throw const NotImplementedError('Google Sign-In callback handling');
  }

  /// Sign in with GitHub
  Future<AuthResult> signInWithGitHub() async {
    final callbackUrl = '$authUrl/api/auth/github/callback';
    final authUrlStr =
        '$authUrl/api/auth/github?redirect_uri=$callbackUrl&project_id=$projectId';

    throw const NotImplementedError('GitHub Sign-In callback handling');
  }

  /// Sign in with Discord
  Future<AuthResult> signInWithDiscord() async {
    final callbackUrl = '$authUrl/api/auth/discord/callback';
    final authUrlStr =
        '$authUrl/api/auth/discord?redirect_uri=$callbackUrl&project_id=$projectId';

    throw const NotImplementedError('Discord Sign-In callback handling');
  }

  /// Sign in with Apple (iOS only)
  Future<AuthResult> signInWithApple() async {
    // Use sign_in_with_apple package on iOS
    throw const NotImplementedError('Apple Sign-In requires platform channel');
  }
}

File: packages/ios-swift/Sources/CinacoinAppKitAuth/SocialLogin.swift
import Foundation
import AuthenticationServices

/// Social login manager for iOS
public final class SocialLoginManager: NSObject, ObservableObject, @unchecked Sendable {
    
    @Published public private(set) var isLoading = false
    @Published public private(set) var error: String?
    
    private let authUrl: String
    private let projectId: String
    
    public init(authUrl: String = "https://auth.cinacoin.com", projectId: String) {
        self.authUrl = authUrl
        self.projectId = projectId
        super.init()
    }
    
    /// Sign in with Apple
    public func signInWithApple(
        presentationAnchor: ASPresentationAnchor? = nil
    ) async throws -> AuthResult {
        isLoading = true
        defer { isLoading = false }
        
        // Request Apple ID credential
        let authorizationController = ASAuthorizationController(
            authorizationRequests: [
                ASAuthorizationAppleIDProvider().createRequest()
            ]
        )
        
        // In production, this would use ASAuthorizationControllerDelegate
        // For now, simulate the flow
        throw AuthError.notImplemented("Apple Sign-In requires ASAuthorizationControllerDelegate")
    }
    
    /// Sign in with Google (via ASWebAuthenticationSession)
    public func signInWithGoogle() async throws -> AuthResult {
        isLoading = true
        defer { isLoading = false }
        
        let callbackURL = "\(authUrl)/api/auth/google/callback"
        let authURL = "\(authUrl)/api/auth/google?redirect_uri=\(callbackURL)&project_id=\(projectId)"
        
        guard let url = URL(string: authURL) else {
            throw AuthError.invalidURL
        }
        
        return try await performWebAuth(url: url, callbackScheme: "cinacoin")
    }
    
    /// Sign in with GitHub
    public func signInWithGitHub() async throws -> AuthResult {
        isLoading = true
        defer { isLoading = false }
        
        let callbackURL = "\(authUrl)/api/auth/github/callback"
        let authURL = "\(authUrl)/api/auth/github?redirect_uri=\(callbackURL)&project_id=\(projectId)"
        
        guard let url = URL(string: authURL) else {
            throw AuthError.invalidURL
        }
        
        return try await performWebAuth(url: url, callbackScheme: "cinacoin")
    }
    
    /// Sign in with Discord
    public func signInWithDiscord() async throws -> AuthResult {
        isLoading = true
        defer { isLoading = false }
        
        let callbackURL = "\(authUrl)/api/auth/discord/callback"
        let authURL = "\(authUrl)/api/auth/discord?redirect_uri=\(callbackURL)&project_id=\(projectId)"
        
        guard let url = URL(string: authURL) else {
            throw AuthError.invalidURL
        }
        
        return try await performWebAuth(url: url, callbackScheme: "cinacoin")
    }
    
    // MARK: - Private
    
    private func performWebAuth(url: URL, callbackScheme: String) async throws -> AuthResult {
        // Web authentication session
        try await withCheckedThrowingContinuation { (continuation: CheckedContinuation<URL, Error>) in
            let session = ASWebAuthenticationSession(
                url: url,
                callbackURLScheme: callbackScheme
            ) { callbackURL, error in
                if let error = error {
                    continuation.resume(throwing: error)
                } else if let callbackURL = callbackURL {
                    continuation.resume(returning: callbackURL)
                } else {
                    continuation.resume(throwing: AuthError.cancelled)
                }
            }
            session.presentationContextProvider = nil
            session.prefersEphemeralWebBrowserSession = true
            session.start()
        }
        
        // Parse callback URL for auth result
        // In production, exchange code for tokens
        throw AuthError.notImplemented("Token exchange not yet implemented")
    }
}

/// Auth errors
public enum AuthError: Error, LocalizedError {
    case invalidURL
    case cancelled
    case notImplemented(String)
    case networkError(String)
    case invalidCredentials
    case tokenExpired
    
    public var errorDescription: String? {
        switch self {
        case .invalidURL: return "Invalid URL"
        case .cancelled: return "Authentication cancelled"
        case .notImplemented(let msg): return "Not implemented: \(msg)"
        case .networkError(let msg): return "Network error: \(msg)"
        case .invalidCredentials: return "Invalid credentials"
        case .tokenExpired: return "Token expired"
        }
    }
}
